src/ir/ssa_names.h

```diff
@@ -0,0 +1,62 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#ifndef SRC_IR_SSA_NAMES_H_
+#define SRC_IR_SSA_NAMES_H_
+
+#include "absl/container/flat_hash_map.h"
+
+namespace raksha::ir {
+
+class Block;
+class Operation;
+
+class SSANames {
+ public:
+  using ID = int;
+  SSANames() : next_block_id_(0), next_operation_id_(0) {}
+
+  // Disable copy (and move) semantics.
+  SSANames(const SSANames &) = delete;
+  SSANames &operator=(const SSANames &) = delete;
+
+  ID GetOrCreateID(const Operation &operation) {
+    return GetOrCreateIDHelper(operation_ids_, next_operation_id_, &operation);
+  }
+
+  ID GetOrCreateID(const Block &block) {
+    return GetOrCreateIDHelper(block_ids_, next_block_id_, &block);
+  }
+
+ private:
+  template <typename T, typename U>
+  static ID GetOrCreateIDHelper(T &ids_table, ID &next_id, U entity) {
+    auto find_result = ids_table.find(entity);
+    if (find_result == ids_table.end()) {
+      ID result = next_id++;
+      ids_table.insert({entity, result});
+      return result;
+    }
+    return find_result->second;
```

Comment:
I think this might be more concisely and idiomatically written like this:

```C++
auto insert_result = ids_table.insert({entity, next_id});
if (insert_result.second) ++next_id;
return insert_result.first->second;
```

`insert` does nothing if a key already exists, returning a `pair` with the iterator to the preexisting entry in `first` and `false` in second. If a key doesn't already exist, it returns the iterator to the new entry in `first` and `true` in second. That means the following:

1. Regardless of whether insertion happened, the ID you want is in the value field of the entry pointed to by the iterator, or `insert_result.first->second`.

2. Whether you need to increment the ID is exactly the same as whether an entry was inserted, or `insert_result.second`.

As an added bonus, you look up in the map once instead of twice, which is a fun optimization, but which probably doesn't matter unless this is in the inner loop somewhere.

Also, preincrement is preferred over postincrement.

Reply:
Done. Used the idiomatic code in `IDManager`.